src/software/new_geom/segment.cpp

```diff
@@ -0,0 +1,59 @@
+#include "software/new_geom/segment.h"
+
+Segment::Segment(const Point& start, const Point& end) : start(start), end(end)
+{
+    if (start == end)
+    {
+        throw std::invalid_argument("Attempting to construct a degenerate segment");
+    }
+}
```

Comment:
I've notice a lot of checks for p1=p2 before constructing a Segment. I feel like many of the use cases of segment involve constructing Segments of unknown length and adding this check everywhere puts a lot of responsibility to the user.

As far as I can tell none of the Segment member functions would fail if p1=p2 - and that case just represents a Segment of length zero.

IMO, p1=p2 should be allowed.

@jonathanlew @akhilveeraghanta

Reply:
It does simplify writing some of these algorithms. I'm not sure if they've been implemented though. Also, we've been moving towards not having degenerate objects in our code base, e.g. World

Reply:
I wouldn't consider a zero-length Segment as degenerate. By that same convention zero magnitude Vectors are degenerate as well

Reply:
A line segment is defined by two distinct points, a segment of length 0 is essentially a point. It makes things easier when using the segment to disallow degenerates imo

Reply:
zero-magnitude vectors have semantic meaning. For example, point1-point1 = Vector(0,0).

Reply:
I agree with Evan that p1=p2 should be allowed. All our current operations on a segment like that are valid, and there's similar semantic meaning to zero-magnitude vectors. Eg. `Segment(p1, p2).length() == 0` when `p1==p2`